xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -671,4 +673,11 @@ boolean isUseProtocolV3() {
       return useProtocolV3;
     }
   }
+
+  interface XdsClientPoolFactory {
+
+    void bootstrap() throws IOException;
```

Comment:
Why adding this method to a Factory class? Why not pass `bootstrapInfo` object to the constructor of `XdsClientPoolFactory` implementation?

Reply:
Ah. This was a very hard API decision, as it was trying to flatten the circular workflow.

The goal was to inject an XdsClient to the resolver. However, creating an XdsClient depends on reading the bootstrap file. Reading bootstrap file does not happen until resolver's `start(...)` is called. So this workflow is circular and we are never able to inject an XdsClient unless we also inject an object that takes responsibility of both reading bootstrap file and creating an XdsClient. This is what an `XdsClientPoolFactory` is 😄 .

Note the layer of `XdsClientFactory` is redundant now. Since we still have the "EDS-only" usage, I will leave it now and clean up in a later PR.

Reply:
The workflow issue can be resolved by passing `bootstrapInfo` as an argument of `newXdsClientObjectPool()` instead. See https://gist.github.com/dapengzhang0/707e4b845fe68648c68d880b76053cb1

`XdsClientPoolFactory` shouldn't do two things: File IO + creating client pool. Also, that would make the factory stateful.

Reply:
Yeah, that works better. Thanks. Updated.

Also, I moved this breaking changes to XdsNameResolver2 so that we can review and merge timely without being blocked by other parts and less risk for server breakages.